AOA/UHID Gamepad Issues.

## Environment
 - Scrcpy 2.7 (prebuilt) on Windows 11.
 - samsung A52s 5G running A14 OneUI.

## Describe the bug

the way the UHID/AOA gamepad shows up messes up the input in some games.

on UHID it shows up as:

- Name: "scrcpy: Xbox One Controller"
- Product: "Vendor_0000_Product_0000"

and on AOA:

- Name: blank, it has no name.
- Product: "Vendor_ffff_Product_ffff"


so far ive had these issues in 2 games, Wuthering Waves and Human Fall Flat:

in human fall flat it straight up doesnt work, the inputs get ignored.
and in wuthering waves the triggers dont work (the game probably sees it as a generic controller).

but if i use LHLaurini's old controller build, it works just fine in both games, probably because it shows up as an actual Xbox controller:

- Name: "Microsoft X-Box 360 Pad"
- Product: "Vendor_045e_Product_028e"



so, what if you emulated an Xbox 360 or an Xbox One controller for Xinput and a PS4 controller for DInput?

after a lot of digging, i found `bool SDL_IsJoystickXInput(SDL_GUID guid)`, idk if its gonna be of any use to you though.

okay so on msys it cant find any of the packages i installed:
```
pacman -S mingw-w64-x86_64-SDL2 \
          mingw-w64-x86_64-ffmpeg \
          mingw-w64-x86_64-libusb

# client build dependencies
pacman -S mingw-w64-x86_64-make \
          mingw-w64-x86_64-gcc \
          mingw-w64-x86_64-pkg-config \
          mingw-w64-x86_64-meson
```

and on the mingw64 shell, i get:

```
Administrator@GHOST-PC MINGW64 ~/scrcpy
# ./release.sh
./gradlew clean

BUILD SUCCESSFUL in 6s
1 actionable task: 1 up-to-date
rm -rf "dist" "build-test" "build-server" \
        "build-win32" "build-win64"
[ -d "build-test" ] || ( mkdir "build-test" && \
        meson setup "build-test" -Db_sanitize=address )
The Meson build system
Version: 1.5.0
Source dir: C:/msys64/home/Administrator/scrcpy
Build dir: C:/msys64/home/Administrator/scrcpy/build-test
Build type: native build
Project name: scrcpy
Project version: 2.7
C compiler for the host machine: cc (gcc 14.1.0 "cc (Rev3, Built by MSYS2 project) 14.1.0")
C linker for the host machine: cc ld.bfd 2.42
Host machine cpu family: x86_64
Host machine cpu: x86_64
Windows resource compiler: GNU windres (GNU Binutils) 2.42
Found pkg-config: YES (C:\msys64\mingw64\bin/pkg-config.EXE) 2.2.0
Run-time dependency libavformat found: YES 60.16.100
Run-time dependency libavcodec found: YES 60.31.102
Run-time dependency libavutil found: YES 58.29.100
Run-time dependency libswresample found: YES 4.12.100
Run-time dependency sdl2 found: YES 2.30.5
Run-time dependency libusb-1.0 found: YES 1.0.27
Library mingw32 found: YES
Library ws2_32 found: YES
Checking for function "strdup" : YES
Checking for function "asprintf" : YES
Checking for function "vasprintf" : YES
Checking for function "nrand48" : NO
Checking for function "jrand48" : NO
Checking for function "reallocarray" : NO
Configuring config.h using configuration
Program ./scripts/build-wrapper.sh found: YES (bash C:/msys64/home/Administrator/scrcpy/server/./scripts/build-wrapper.sh)
Build targets in project: 14

scrcpy 2.7

  User defined options
    b_sanitize: address

Found ninja-1.12.1 at C:\msys64\mingw64\bin/ninja.EXE
ninja -C "build-test"
ninja: Entering directory `build-test'
[71/128] Compiling C object app/test_adb_parser.exe.p/tests_test_adb_parser.c.obj
<command-line>: warning: no previous prototype for 'SDL_main' [-Wmissing-prototypes]
../app/tests/test_adb_parser.c:256:5: note: in expansion of macro 'main'
  256 | int main(int argc, char *argv[]) {
      |     ^~~~
[77/128] Compiling C object app/test_binary.exe.p/tests_test_binary.c.obj
<command-line>: warning: no previous prototype for 'SDL_main' [-Wmissing-prototypes]
../app/tests/test_binary.c:138:5: note: in expansion of macro 'main'
  138 | int main(int argc, char *argv[]) {
      |     ^~~~
[80/128] Compiling C object app/test_audiobuf.exe.p/tests_test_audiobuf.c.obj
<command-line>: warning: no previous prototype for 'SDL_main' [-Wmissing-prototypes]
../app/tests/test_audiobuf.c:119:5: note: in expansion of macro 'main'
  119 | int main(int argc, char *argv[]) {
      |     ^~~~
[84/128] Compiling C object app/test_cli.exe.p/tests_test_cli.c.obj
<command-line>: warning: no previous prototype for 'SDL_main' [-Wmissing-prototypes]
../app/tests/test_cli.c:154:5: note: in expansion of macro 'main'
  154 | int main(int argc, char *argv[]) {
      |     ^~~~
[91/128] Compiling C object app/test_control_msg...ize.exe.p/tests_test_control_msg_serialize.c.obj
<command-line>: warning: no previous prototype for 'SDL_main' [-Wmissing-prototypes]
../app/tests/test_control_msg_serialize.c:410:5: note: in expansion of macro 'main'
  410 | int main(int argc, char *argv[]) {
      |     ^~~~
[97/128] Compiling C object app/test_device_msg_...ze.exe.p/tests_test_device_msg_deserialize.c.obj
<command-line>: warning: no previous prototype for 'SDL_main' [-Wmissing-prototypes]
../app/tests/test_device_msg_deserialize.c:86:5: note: in expansion of macro 'main'
   86 | int main(int argc, char *argv[]) {
      |     ^~~~
[101/128] Compiling C object app/test_orientation.exe.p/tests_test_orientation.c.obj
<command-line>: warning: no previous prototype for 'SDL_main' [-Wmissing-prototypes]
../app/tests/test_orientation.c:85:5: note: in expansion of macro 'main'
   85 | int main(int argc, char *argv[]) {
      |     ^~~~
[103/128] Compiling C object app/test_strbuf.exe.p/tests_test_strbuf.c.obj
<command-line>: warning: no previous prototype for 'SDL_main' [-Wmissing-prototypes]
../app/tests/test_strbuf.c:42:5: note: in expansion of macro 'main'
   42 | int main(int argc, char *argv[]) {
      |     ^~~~
[107/128] Compiling C object app/test_str.exe.p/tests_test_str.c.obj
<command-line>: warning: no previous prototype for 'SDL_main' [-Wmissing-prototypes]
../app/tests/test_str.c:379:5: note: in expansion of macro 'main'
  379 | int main(int argc, char *argv[]) {
      |     ^~~~
[110/128] Compiling C object app/test_vecdeque.exe.p/tests_test_vecdeque.c.obj
<command-line>: warning: no previous prototype for 'SDL_main' [-Wmissing-prototypes]
../app/tests/test_vecdeque.c:187:5: note: in expansion of macro 'main'
  187 | int main(int argc, char *argv[]) {
      |     ^~~~
[114/128] Compiling C object app/test_vector.exe.p/tests_test_vector.c.obj
<command-line>: warning: no previous prototype for 'SDL_main' [-Wmissing-prototypes]
../app/tests/test_vector.c:403:5: note: in expansion of macro 'main'
  403 | int main(int argc, char *argv[]) {
      |     ^~~~
[116/128] Linking target app/test_adb_parser.exe
FAILED: app/test_adb_parser.exe
"cc"  -o app/test_adb_parser.exe app/test_adb_parser.exe.p/tests_test_adb_parser.c.obj app/test_adb_parser.exe.p/src_adb_adb_device.c.obj app/test_adb_parser.exe.p/src_adb_adb_parser.c.obj app/test_adb_parser.exe.p/src_util_str.c.obj app/test_adb_parser.exe.p/src_util_strbuf.c.obj app/test_adb_parser.exe.p/src_compat.c.obj "-fsanitize=address" "-Wl,--allow-shlib-undefined" "-Wl,--start-group" "C:/msys64/mingw64/lib/libavformat.dll.a" "C:/msys64/mingw64/lib/libavcodec.dll.a" "C:/msys64/mingw64/lib/libavutil.dll.a" "C:/msys64/mingw64/lib/libswresample.dll.a" "C:/msys64/mingw64/lib/libmingw32.a" "-mwindows" "C:/msys64/mingw64/lib/libSDL2main.a" "C:/msys64/mingw64/lib/libSDL2.dll.a" "C:/msys64/mingw64/lib/libusb-1.0.dll.a" "-lmingw32" "-lws2_32" "-Wl,--subsystem,console" "-lkernel32" "-luser32" "-lgdi32" "-lwinspool" "-lshell32" "-lole32" "-loleaut32" "-luuid" "-lcomdlg32" "-ladvapi32" "-Wl,--end-group"
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lasan: No such file or directory
collect2.exe: error: ld returned 1 exit status
[117/128] Linking target app/scrcpy.exe
FAILED: app/scrcpy.exe
"cc"  -o app/scrcpy.exe app/app_scrcpy-windows.rc_scrcpy-windows.o app/scrcpy.exe.p/src_main.c.obj app/scrcpy.exe.p/src_adb_adb.c.obj app/scrcpy.exe.p/src_adb_adb_device.c.obj app/scrcpy.exe.p/src_adb_adb_parser.c.obj app/scrcpy.exe.p/src_adb_adb_tunnel.c.obj app/scrcpy.exe.p/src_audio_player.c.obj app/scrcpy.exe.p/src_cli.c.obj app/scrcpy.exe.p/src_clock.c.obj app/scrcpy.exe.p/src_compat.c.obj app/scrcpy.exe.p/src_control_msg.c.obj app/scrcpy.exe.p/src_controller.c.obj app/scrcpy.exe.p/src_decoder.c.obj app/scrcpy.exe.p/src_delay_buffer.c.obj app/scrcpy.exe.p/src_demuxer.c.obj app/scrcpy.exe.p/src_device_msg.c.obj app/scrcpy.exe.p/src_display.c.obj app/scrcpy.exe.p/src_events.c.obj app/scrcpy.exe.p/src_icon.c.obj app/scrcpy.exe.p/src_file_pusher.c.obj app/scrcpy.exe.p/src_fps_counter.c.obj app/scrcpy.exe.p/src_frame_buffer.c.obj app/scrcpy.exe.p/src_input_manager.c.obj app/scrcpy.exe.p/src_keyboard_sdk.c.obj app/scrcpy.exe.p/src_mouse_sdk.c.obj app/scrcpy.exe.p/src_opengl.c.obj app/scrcpy.exe.p/src_options.c.obj app/scrcpy.exe.p/src_packet_merger.c.obj app/scrcpy.exe.p/src_receiver.c.obj app/scrcpy.exe.p/src_recorder.c.obj app/scrcpy.exe.p/src_scrcpy.c.obj app/scrcpy.exe.p/src_screen.c.obj app/scrcpy.exe.p/src_server.c.obj app/scrcpy.exe.p/src_version.c.obj app/scrcpy.exe.p/src_hid_hid_gamepad.c.obj app/scrcpy.exe.p/src_hid_hid_keyboard.c.obj app/scrcpy.exe.p/src_hid_hid_mouse.c.obj app/scrcpy.exe.p/src_trait_frame_source.c.obj app/scrcpy.exe.p/src_trait_packet_source.c.obj app/scrcpy.exe.p/src_uhid_gamepad_uhid.c.obj app/scrcpy.exe.p/src_uhid_keyboard_uhid.c.obj app/scrcpy.exe.p/src_uhid_mouse_uhid.c.obj app/scrcpy.exe.p/src_uhid_uhid_output.c.obj app/scrcpy.exe.p/src_util_acksync.c.obj app/scrcpy.exe.p/src_util_audiobuf.c.obj app/scrcpy.exe.p/src_util_average.c.obj app/scrcpy.exe.p/src_util_file.c.obj app/scrcpy.exe.p/src_util_intmap.c.obj app/scrcpy.exe.p/src_util_intr.c.obj app/scrcpy.exe.p/src_util_log.c.obj app/scrcpy.exe.p/src_util_memory.c.obj app/scrcpy.exe.p/src_util_net.c.obj app/scrcpy.exe.p/src_util_net_intr.c.obj app/scrcpy.exe.p/src_util_process.c.obj app/scrcpy.exe.p/src_util_process_intr.c.obj app/scrcpy.exe.p/src_util_rand.c.obj app/scrcpy.exe.p/src_util_strbuf.c.obj app/scrcpy.exe.p/src_util_str.c.obj app/scrcpy.exe.p/src_util_term.c.obj app/scrcpy.exe.p/src_util_thread.c.obj app/scrcpy.exe.p/src_util_tick.c.obj app/scrcpy.exe.p/src_util_timeout.c.obj app/scrcpy.exe.p/src_sys_win_file.c.obj app/scrcpy.exe.p/src_sys_win_process.c.obj app/scrcpy.exe.p/src_usb_aoa_hid.c.obj app/scrcpy.exe.p/src_usb_gamepad_aoa.c.obj app/scrcpy.exe.p/src_usb_keyboard_aoa.c.obj app/scrcpy.exe.p/src_usb_mouse_aoa.c.obj app/scrcpy.exe.p/src_usb_scrcpy_otg.c.obj app/scrcpy.exe.p/src_usb_screen_otg.c.obj app/scrcpy.exe.p/src_usb_usb.c.obj "-fsanitize=address" "-Wl,--allow-shlib-undefined" "-Wl,--start-group" "C:/msys64/mingw64/lib/libavformat.dll.a" "C:/msys64/mingw64/lib/libavcodec.dll.a" "C:/msys64/mingw64/lib/libavutil.dll.a" "C:/msys64/mingw64/lib/libswresample.dll.a" "C:/msys64/mingw64/lib/libmingw32.a" "-mwindows" "C:/msys64/mingw64/lib/libSDL2main.a" "C:/msys64/mingw64/lib/libSDL2.dll.a" "C:/msys64/mingw64/lib/libusb-1.0.dll.a" "-lmingw32" "-lws2_32" "-Wl,--subsystem,console" "-lkernel32" "-luser32" "-lgdi32" "-lwinspool" "-lshell32" "-lole32" "-loleaut32" "-luuid" "-lcomdlg32" "-ladvapi32" "-Wl,--end-group"
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lasan: No such file or directory
collect2.exe: error: ld returned 1 exit status
[118/128] Linking target app/test_binary.exe
FAILED: app/test_binary.exe
"cc"  -o app/test_binary.exe app/test_binary.exe.p/tests_test_binary.c.obj app/test_binary.exe.p/src_compat.c.obj "-fsanitize=address" "-Wl,--allow-shlib-undefined" "-Wl,--start-group" "C:/msys64/mingw64/lib/libavformat.dll.a" "C:/msys64/mingw64/lib/libavcodec.dll.a" "C:/msys64/mingw64/lib/libavutil.dll.a" "C:/msys64/mingw64/lib/libswresample.dll.a" "C:/msys64/mingw64/lib/libmingw32.a" "-mwindows" "C:/msys64/mingw64/lib/libSDL2main.a" "C:/msys64/mingw64/lib/libSDL2.dll.a" "C:/msys64/mingw64/lib/libusb-1.0.dll.a" "-lmingw32" "-lws2_32" "-Wl,--subsystem,console" "-lkernel32" "-luser32" "-lgdi32" "-lwinspool" "-lshell32" "-lole32" "-loleaut32" "-luuid" "-lcomdlg32" "-ladvapi32" "-Wl,--end-group"
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/14.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lasan: No such file or directory
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
make: *** [release.mk:56: test] Error 1
```

`./release.sh` is the script to build the whole releases from Linux. That's not what you want.

You just want to build natively, so once you installed the packages, follow [these steps](https://github.com/Genymobile/scrcpy/blob/master/doc/build.md#build).

> ```
> cannot find -lasan: No such file or directory
> ```

The tests are built with ASAN enabled in the release script, but ASAN is not available on Windows. That won't happen if you build manually (without enabling ASAN explicitly). https://github.com/Genymobile/scrcpy/issues/4736#issuecomment-1983443269

> warning: no previous prototype for 'SDL_main'

Oh, I'll have to fix this SDL warning again…

i got
```
app/deps/sdl.sh: line 47: x86_64-w64-mingw32-strip: command not found
make: *** [release64.mk:54: prepare-deps-win64] Error 127
```
ig i have to replace `${HOST_TRIPLET}-strip "$INSTALL_DIR/$HOST/bin/SDL2.dll"` in sdl.sh with `strip "$INSTALL_DIR/$HOST/bin/SDL2.dll"`?

--- a/app/src/control_msg.c
+++ b/app/src/control_msg.c
@@ -152,8 +152,10 @@ sc_control_msg_serialize(const struct sc_control_msg *msg, uint8_t *buf) {
             return 2;
         case SC_CONTROL_MSG_TYPE_UHID_CREATE:
             sc_write16be(&buf[1], msg->uhid_create.id);
+            sc_write16be(&buf[3], msg->uhid_create.vendor_id);
+            sc_write16be(&buf[5], msg->uhid_create.product_id);
 
-            size_t index = 3;
+            size_t index = 7;
             index += write_string_tiny(&buf[index], msg->uhid_create.name, 127);
 
             sc_write16be(&buf[index], msg->uhid_create.report_desc_size);
@@ -278,9 +280,13 @@ sc_control_msg_log(const struct sc_control_msg *msg) {
             // Quote only if name is not null
             const char *name = msg->uhid_create.name;
             const char *quote = name ? "\"" : "";
-            LOG_CMSG("UHID create [%" PRIu16 "] name=%s%s%s "
-                     "report_desc_size=%" PRIu16, msg->uhid_create.id,
-                     quote, name, quote, msg->uhid_create.report_desc_size);
+            LOG_CMSG("UHID create [%" PRIu16 "] %04" PRIx16 ":%04" PRIx16
+                     " name=%s%s%s report_desc_size=%" PRIu16,
+                     msg->uhid_create.id,
+                     msg->uhid_create.vendor_id,
+                     msg->uhid_create.product_id,
+                     quote, name, quote,
+                     msg->uhid_create.report_desc_size);
             break;
         }
         case SC_CONTROL_MSG_TYPE_UHID_INPUT: {

--- a/app/src/control_msg.h
+++ b/app/src/control_msg.h
@@ -94,6 +94,8 @@ struct sc_control_msg {
         } set_display_power;
         struct {
             uint16_t id;
+            uint16_t vendor_id;
+            uint16_t product_id;
             const char *name; // pointer to static data
             uint16_t report_desc_size;
             const uint8_t *report_desc; // pointer to static data

--- a/app/src/hid/hid_event.h
+++ b/app/src/hid/hid_event.h
@@ -15,7 +15,6 @@ struct sc_hid_input {
 
 struct sc_hid_open {
     uint16_t hid_id;
-    const char *name; // pointer to static memory
     const uint8_t *report_desc; // pointer to static memory
     size_t report_desc_size;
 };

--- a/app/src/hid/hid_gamepad.c
+++ b/app/src/hid/hid_gamepad.c
@@ -246,14 +246,8 @@ sc_hid_gamepad_generate_open(struct sc_hid_gamepad *hid,
 
     sc_hid_gamepad_slot_init(&hid->slots[slot_idx], gamepad_id);
 
-    SDL_GameController* game_controller =
-        SDL_GameControllerFromInstanceID(gamepad_id);
-    assert(game_controller);
-    const char *name = SDL_GameControllerName(game_controller);
-
     uint16_t hid_id = sc_hid_gamepad_slot_get_id(slot_idx);
     hid_open->hid_id = hid_id;
-    hid_open->name = name;
     hid_open->report_desc = SC_HID_GAMEPAD_REPORT_DESC;
     hid_open->report_desc_size = sizeof(SC_HID_GAMEPAD_REPORT_DESC);
 

--- a/app/src/hid/hid_gamepad.c
+++ b/app/src/hid/hid_gamepad.c
@@ -65,7 +65,7 @@ static const uint8_t SC_HID_GAMEPAD_REPORT_DESC[] = {
     0x75, 0x10,
     // Report Count (4)
     0x95, 0x04,
-    // Input (Data, Variable, Absolute): 4 bytes (X, Y, Z, Rz)
+    // Input (Data, Variable, Absolute): 4x2 bytes (X, Y, Z, Rz)
     0x81, 0x02,
 
     // Usage Page (Simulation Controls)
@@ -82,7 +82,7 @@ static const uint8_t SC_HID_GAMEPAD_REPORT_DESC[] = {
     0x75, 0x10,
     // Report Count (2)
     0x95, 0x02,
-    // Input (Data, Variable, Absolute): 2 bytes (L2, R2)
+    // Input (Data, Variable, Absolute): 2x2 bytes (L2, R2)
     0x81, 0x02,
 
     // Usage Page (Buttons)
@@ -182,7 +182,7 @@ static const uint8_t SC_HID_GAMEPAD_REPORT_DESC[] = {
  *                     `------------- SC_GAMEPAD_BUTTON_RIGHT_STICK
  *
  *                  +---------------+
- *        byte 14:  |0 0 0 . . . . .| hat switch (dpad) position (0-8)
+ *        byte 14:  |0 0 0 0 . . . .| hat switch (dpad) position (0-8)
  *                  +---------------+
  *                     9 possible positions and their values:
  *                             8 1 2

--- a/app/src/hid/hid_gamepad.c
+++ b/app/src/hid/hid_gamepad.c
@@ -52,10 +52,10 @@ static const uint8_t SC_HID_GAMEPAD_REPORT_DESC[] = {
     0x09, 0x30,
     // Usage (Y)   Left stick y
     0x09, 0x31,
-    // Usage (Z)   Right stick x
-    0x09, 0x32,
-    // Usage (Rz)  Right stick y
-    0x09, 0x35,
+    // Usage (Rx)  Right stick x
+    0x09, 0x33,
+    // Usage (Ry)  Right stick y
+    0x09, 0x34,
     // Logical Minimum (0)
     0x15, 0x00,
     // Logical Maximum (65535)
@@ -68,12 +68,12 @@ static const uint8_t SC_HID_GAMEPAD_REPORT_DESC[] = {
     // Input (Data, Variable, Absolute): 4x2 bytes (X, Y, Z, Rz)
     0x81, 0x02,
 
-    // Usage Page (Simulation Controls)
-    0x05, 0x02,
-    // Usage (Brake)
-    0x09, 0xC5,
-    // Usage (Accelerator)
-    0x09, 0xC4,
+    // Usage Page (Generic Desktop)
+    0x05, 0x01,
+    // Usage (Z)
+    0x09, 0x32,
+    // Usage (Rz)
+    0x09, 0x35,
     // Logical Minimum (0)
     0x15, 0x00,
     // Logical Maximum (32767)

--- a/app/src/hid/hid_gamepad.c
+++ b/app/src/hid/hid_gamepad.c
@@ -191,16 +191,19 @@ static const uint8_t SC_HID_GAMEPAD_REPORT_DESC[] = {
  *                     (8 is top-left, 1 is top, 2 is top-right, etc.)
  */
 
+// [-32768 to 32767] -> [0 to 65535]
+#define AXIS_RESCALE(V) (uint16_t) (((int32_t) V) + 0x8000)
+
 static void
 sc_hid_gamepad_slot_init(struct sc_hid_gamepad_slot *slot,
                          uint32_t gamepad_id) {
     assert(gamepad_id != SC_GAMEPAD_ID_INVALID);
     slot->gamepad_id = gamepad_id;
     slot->buttons = 0;
-    slot->axis_left_x = 0;
-    slot->axis_left_y = 0;
-    slot->axis_right_x = 0;
-    slot->axis_right_y = 0;
+    slot->axis_left_x = AXIS_RESCALE(0);
+    slot->axis_left_y = AXIS_RESCALE(0);
+    slot->axis_right_x = AXIS_RESCALE(0);
+    slot->axis_right_y = AXIS_RESCALE(0);
     slot->axis_left_trigger = 0;
     slot->axis_right_trigger = 0;
 }
@@ -423,8 +426,6 @@ sc_hid_gamepad_generate_input_from_axis(struct sc_hid_gamepad *hid,
 
     struct sc_hid_gamepad_slot *slot = &hid->slots[slot_idx];
 
-// [-32768 to 32767] -> [0 to 65535]
-#define AXIS_RESCALE(V) (uint16_t) (((int32_t) V) + 0x8000)
     switch (event->axis) {
         case SC_GAMEPAD_AXIS_LEFTX:
             slot->axis_left_x = AXIS_RESCALE(event->value);

--- a/app/src/hid/hid_keyboard.c
+++ b/app/src/hid/hid_keyboard.c
@@ -335,7 +335,6 @@ sc_hid_keyboard_generate_input_from_mods(struct sc_hid_input *hid_input,
 
 void sc_hid_keyboard_generate_open(struct sc_hid_open *hid_open) {
     hid_open->hid_id = SC_HID_ID_KEYBOARD;
-    hid_open->name = NULL; // No name specified after "scrcpy"
     hid_open->report_desc = SC_HID_KEYBOARD_REPORT_DESC;
     hid_open->report_desc_size = sizeof(SC_HID_KEYBOARD_REPORT_DESC);
 }

--- a/app/src/hid/hid_mouse.c
+++ b/app/src/hid/hid_mouse.c
@@ -190,7 +190,6 @@ sc_hid_mouse_generate_input_from_scroll(struct sc_hid_input *hid_input,
 
 void sc_hid_mouse_generate_open(struct sc_hid_open *hid_open) {
     hid_open->hid_id = SC_HID_ID_MOUSE;
-    hid_open->name = NULL; // No name specified after "scrcpy"
     hid_open->report_desc = SC_HID_MOUSE_REPORT_DESC;
     hid_open->report_desc_size = sizeof(SC_HID_MOUSE_REPORT_DESC);
 }

--- a/app/src/input_events.h
+++ b/app/src/input_events.h
@@ -412,18 +412,12 @@ struct sc_touch_event {
     float pressure;
 };
 
-enum sc_gamepad_device_event_type {
-    SC_GAMEPAD_DEVICE_ADDED,
-    SC_GAMEPAD_DEVICE_REMOVED,
-};
-
 // As documented in <https://wiki.libsdl.org/SDL2/SDL_JoystickID>:
 // The ID value starts at 0 and increments from there. The value -1 is an
 // invalid ID.
 #define SC_GAMEPAD_ID_INVALID UINT32_C(-1)
 
 struct sc_gamepad_device_event {
-    enum sc_gamepad_device_event_type type;
     uint32_t gamepad_id;
 };
 
@@ -501,16 +495,6 @@ sc_mouse_buttons_state_from_sdl(uint32_t buttons_state) {
 
     // SC_MOUSE_BUTTON_* constants are initialized from SDL_BUTTON(index)
     return buttons_state;
-}
-
-static inline enum sc_gamepad_device_event_type
-sc_gamepad_device_event_type_from_sdl_type(uint32_t type) {
-    assert(type == SDL_CONTROLLERDEVICEADDED
-        || type == SDL_CONTROLLERDEVICEREMOVED);
-    if (type == SDL_CONTROLLERDEVICEADDED) {
-        return SC_GAMEPAD_DEVICE_ADDED;
-    }
-    return SC_GAMEPAD_DEVICE_REMOVED;
 }
 
 static inline enum sc_gamepad_axis

--- a/app/src/input_manager.c
+++ b/app/src/input_manager.c
@@ -908,7 +908,6 @@ sc_input_manager_process_mouse_wheel(struct sc_input_manager *im,
 static void
 sc_input_manager_process_gamepad_device(struct sc_input_manager *im,
                                        const SDL_ControllerDeviceEvent *event) {
-    SDL_JoystickID id;
     if (event->type == SDL_CONTROLLERDEVICEADDED) {
         SDL_GameController *gc = SDL_GameControllerOpen(event->which);
         if (!gc) {
@@ -923,9 +922,12 @@ sc_input_manager_process_gamepad_device(struct sc_input_manager *im,
             return;
         }
 
-        id = SDL_JoystickInstanceID(joystick);
+        struct sc_gamepad_device_event evt = {
+            .gamepad_id = SDL_JoystickInstanceID(joystick),
+        };
+        im->gp->ops->process_gamepad_added(im->gp, &evt);
     } else if (event->type == SDL_CONTROLLERDEVICEREMOVED) {
-        id = event->which;
+        SDL_JoystickID id = event->which;
 
         SDL_GameController *gc = SDL_GameControllerFromInstanceID(id);
         if (gc) {
@@ -933,16 +935,15 @@ sc_input_manager_process_gamepad_device(struct sc_input_manager *im,
         } else {
             LOGW("Unknown gamepad device removed");
         }
+
+        struct sc_gamepad_device_event evt = {
+            .gamepad_id = id,
+        };
+        im->gp->ops->process_gamepad_removed(im->gp, &evt);
     } else {
         // Nothing to do
         return;
     }
-
-    struct sc_gamepad_device_event evt = {
-        .type = sc_gamepad_device_event_type_from_sdl_type(event->type),
-        .gamepad_id = id,
-    };
-    im->gp->ops->process_gamepad_device(im->gp, &evt);
 }
 
 static void

--- a/app/src/trait/gamepad_processor.h
+++ b/app/src/trait/gamepad_processor.h
@@ -20,13 +20,22 @@ struct sc_gamepad_processor {
 struct sc_gamepad_processor_ops {
 
     /**
-     * Process a gamepad device added or removed
+     * Process a gamepad device added event
      *
      * This function is mandatory.
      */
     void
-    (*process_gamepad_device)(struct sc_gamepad_processor *gp,
-                              const struct sc_gamepad_device_event *event);
+    (*process_gamepad_added)(struct sc_gamepad_processor *gp,
+                             const struct sc_gamepad_device_event *event);
+
+    /**
+     * Process a gamepad device removed event
+     *
+     * This function is mandatory.
+     */
+    void
+    (*process_gamepad_removed)(struct sc_gamepad_processor *gp,
+                               const struct sc_gamepad_device_event *event);
 
     /**
      * Process a gamepad axis event

--- a/app/src/uhid/gamepad_uhid.c
+++ b/app/src/uhid/gamepad_uhid.c
@@ -7,8 +7,9 @@
 /** Downcast gamepad processor to sc_gamepad_uhid */
 #define DOWNCAST(GP) container_of(GP, struct sc_gamepad_uhid, gamepad_processor)
 
-#define SC_GAMEPAD_UHID_VENDOR_ID 0
-#define SC_GAMEPAD_UHID_PRODUCT_ID 0
+// Xbox 360
+#define SC_GAMEPAD_UHID_VENDOR_ID UINT16_C(0x045e)
+#define SC_GAMEPAD_UHID_PRODUCT_ID UINT16_C(0x028e)
 
 static void
 sc_gamepad_uhid_send_input(struct sc_gamepad_uhid *gamepad,

--- a/app/src/uhid/gamepad_uhid.c
+++ b/app/src/uhid/gamepad_uhid.c
@@ -7,6 +7,9 @@
 /** Downcast gamepad processor to sc_gamepad_uhid */
 #define DOWNCAST(GP) container_of(GP, struct sc_gamepad_uhid, gamepad_processor)
 
+#define SC_GAMEPAD_UHID_VENDOR_ID 0
+#define SC_GAMEPAD_UHID_PRODUCT_ID 0
+
 static void
 sc_gamepad_uhid_send_input(struct sc_gamepad_uhid *gamepad,
                            const struct sc_hid_input *hid_input,
@@ -30,6 +33,8 @@ sc_gamepad_uhid_send_open(struct sc_gamepad_uhid *gamepad,
     struct sc_control_msg msg;
     msg.type = SC_CONTROL_MSG_TYPE_UHID_CREATE;
     msg.uhid_create.id = hid_open->hid_id;
+    msg.uhid_create.vendor_id = SC_GAMEPAD_UHID_VENDOR_ID;
+    msg.uhid_create.product_id = SC_GAMEPAD_UHID_PRODUCT_ID;
     msg.uhid_create.name = hid_open->name;
     msg.uhid_create.report_desc = hid_open->report_desc;
     msg.uhid_create.report_desc_size = hid_open->report_desc_size;

--- a/app/src/uhid/gamepad_uhid.c
+++ b/app/src/uhid/gamepad_uhid.c
@@ -52,29 +52,31 @@ sc_gamepad_uhid_send_close(struct sc_gamepad_uhid *gamepad,
 }
 
 static void
-sc_gamepad_processor_process_gamepad_device(struct sc_gamepad_processor *gp,
+sc_gamepad_processor_process_gamepad_added(struct sc_gamepad_processor *gp,
                                 const struct sc_gamepad_device_event *event) {
     struct sc_gamepad_uhid *gamepad = DOWNCAST(gp);
 
-    if (event->type == SC_GAMEPAD_DEVICE_ADDED) {
-        struct sc_hid_open hid_open;
-        if (!sc_hid_gamepad_generate_open(&gamepad->hid, &hid_open,
-                                          event->gamepad_id)) {
-            return;
-        }
-
-        sc_gamepad_uhid_send_open(gamepad, &hid_open);
-    } else {
-        assert(event->type == SC_GAMEPAD_DEVICE_REMOVED);
-
-        struct sc_hid_close hid_close;
-        if (!sc_hid_gamepad_generate_close(&gamepad->hid, &hid_close,
-                                           event->gamepad_id)) {
-            return;
-        }
-
-        sc_gamepad_uhid_send_close(gamepad, &hid_close);
+    struct sc_hid_open hid_open;
+    if (!sc_hid_gamepad_generate_open(&gamepad->hid, &hid_open,
+                                      event->gamepad_id)) {
+        return;
     }
+
+    sc_gamepad_uhid_send_open(gamepad, &hid_open);
+}
+
+static void
+sc_gamepad_processor_process_gamepad_removed(struct sc_gamepad_processor *gp,
+                                const struct sc_gamepad_device_event *event) {
+    struct sc_gamepad_uhid *gamepad = DOWNCAST(gp);
+
+    struct sc_hid_close hid_close;
+    if (!sc_hid_gamepad_generate_close(&gamepad->hid, &hid_close,
+                                       event->gamepad_id)) {
+        return;
+    }
+
+    sc_gamepad_uhid_send_close(gamepad, &hid_close);
 }
 
 static void
@@ -114,7 +116,8 @@ sc_gamepad_uhid_init(struct sc_gamepad_uhid *gamepad,
     gamepad->controller = controller;
 
     static const struct sc_gamepad_processor_ops ops = {
-        .process_gamepad_device = sc_gamepad_processor_process_gamepad_device,
+        .process_gamepad_added = sc_gamepad_processor_process_gamepad_added,
+        .process_gamepad_removed = sc_gamepad_processor_process_gamepad_removed,
         .process_gamepad_axis = sc_gamepad_processor_process_gamepad_axis,
         .process_gamepad_button = sc_gamepad_processor_process_gamepad_button,
     };

--- a/app/src/uhid/gamepad_uhid.c
+++ b/app/src/uhid/gamepad_uhid.c
@@ -69,6 +69,12 @@ sc_gamepad_processor_process_gamepad_added(struct sc_gamepad_processor *gp,
         return;
     }
 
+    SDL_GameController* game_controller =
+        SDL_GameControllerFromInstanceID(event->gamepad_id);
+    assert(game_controller);
+    const char *name = SDL_GameControllerName(game_controller);
+    LOGI("Gamepad added: [%" PRIu32 "] %s", event->gamepad_id, name);
+
     sc_gamepad_uhid_send_open(gamepad, &hid_open);
 }
 
@@ -82,6 +88,8 @@ sc_gamepad_processor_process_gamepad_removed(struct sc_gamepad_processor *gp,
                                        event->gamepad_id)) {
         return;
     }
+
+    LOGI("Gamepad removed: [%" PRIu32 "]", event->gamepad_id);
 
     sc_gamepad_uhid_send_close(gamepad, &hid_close);
 }

--- a/app/src/uhid/gamepad_uhid.c
+++ b/app/src/uhid/gamepad_uhid.c
@@ -10,6 +10,7 @@
 // Xbox 360
 #define SC_GAMEPAD_UHID_VENDOR_ID UINT16_C(0x045e)
 #define SC_GAMEPAD_UHID_PRODUCT_ID UINT16_C(0x028e)
+#define SC_GAMEPAD_UHID_NAME "Microsoft X-Box 360 Pad"
 
 static void
 sc_gamepad_uhid_send_input(struct sc_gamepad_uhid *gamepad,
@@ -36,7 +37,7 @@ sc_gamepad_uhid_send_open(struct sc_gamepad_uhid *gamepad,
     msg.uhid_create.id = hid_open->hid_id;
     msg.uhid_create.vendor_id = SC_GAMEPAD_UHID_VENDOR_ID;
     msg.uhid_create.product_id = SC_GAMEPAD_UHID_PRODUCT_ID;
-    msg.uhid_create.name = hid_open->name;
+    msg.uhid_create.name = SC_GAMEPAD_UHID_NAME;
     msg.uhid_create.report_desc = hid_open->report_desc;
     msg.uhid_create.report_desc_size = hid_open->report_desc_size;
 

--- a/app/src/uhid/keyboard_uhid.c
+++ b/app/src/uhid/keyboard_uhid.c
@@ -143,7 +143,7 @@ sc_keyboard_uhid_init(struct sc_keyboard_uhid *kb,
     msg.uhid_create.id = SC_HID_ID_KEYBOARD;
     msg.uhid_create.vendor_id = 0;
     msg.uhid_create.product_id = 0;
-    msg.uhid_create.name = hid_open.name;
+    msg.uhid_create.name = NULL;
     msg.uhid_create.report_desc = hid_open.report_desc;
     msg.uhid_create.report_desc_size = hid_open.report_desc_size;
     if (!sc_controller_push_msg(controller, &msg)) {

--- a/app/src/uhid/keyboard_uhid.c
+++ b/app/src/uhid/keyboard_uhid.c
@@ -141,6 +141,8 @@ sc_keyboard_uhid_init(struct sc_keyboard_uhid *kb,
     struct sc_control_msg msg;
     msg.type = SC_CONTROL_MSG_TYPE_UHID_CREATE;
     msg.uhid_create.id = SC_HID_ID_KEYBOARD;
+    msg.uhid_create.vendor_id = 0;
+    msg.uhid_create.product_id = 0;
     msg.uhid_create.name = hid_open.name;
     msg.uhid_create.report_desc = hid_open.report_desc;
     msg.uhid_create.report_desc_size = hid_open.report_desc_size;

--- a/app/src/uhid/mouse_uhid.c
+++ b/app/src/uhid/mouse_uhid.c
@@ -81,6 +81,8 @@ sc_mouse_uhid_init(struct sc_mouse_uhid *mouse,
     struct sc_control_msg msg;
     msg.type = SC_CONTROL_MSG_TYPE_UHID_CREATE;
     msg.uhid_create.id = SC_HID_ID_MOUSE;
+    msg.uhid_create.vendor_id = 0;
+    msg.uhid_create.product_id = 0;
     msg.uhid_create.name = hid_open.name;
     msg.uhid_create.report_desc = hid_open.report_desc;
     msg.uhid_create.report_desc_size = hid_open.report_desc_size;

--- a/app/src/uhid/mouse_uhid.c
+++ b/app/src/uhid/mouse_uhid.c
@@ -83,7 +83,7 @@ sc_mouse_uhid_init(struct sc_mouse_uhid *mouse,
     msg.uhid_create.id = SC_HID_ID_MOUSE;
     msg.uhid_create.vendor_id = 0;
     msg.uhid_create.product_id = 0;
-    msg.uhid_create.name = hid_open.name;
+    msg.uhid_create.name = NULL;
     msg.uhid_create.report_desc = hid_open.report_desc;
     msg.uhid_create.report_desc_size = hid_open.report_desc_size;
     if (!sc_controller_push_msg(controller, &msg)) {

--- a/app/src/usb/gamepad_aoa.c
+++ b/app/src/usb/gamepad_aoa.c
@@ -7,33 +7,35 @@
 #define DOWNCAST(GP) container_of(GP, struct sc_gamepad_aoa, gamepad_processor)
 
 static void
-sc_gamepad_processor_process_gamepad_device(struct sc_gamepad_processor *gp,
+sc_gamepad_processor_process_gamepad_added(struct sc_gamepad_processor *gp,
                                 const struct sc_gamepad_device_event *event) {
     struct sc_gamepad_aoa *gamepad = DOWNCAST(gp);
 
-    if (event->type == SC_GAMEPAD_DEVICE_ADDED) {
-        struct sc_hid_open hid_open;
-        if (!sc_hid_gamepad_generate_open(&gamepad->hid, &hid_open,
-                                          event->gamepad_id)) {
-            return;
-        }
+    struct sc_hid_open hid_open;
+    if (!sc_hid_gamepad_generate_open(&gamepad->hid, &hid_open,
+                                      event->gamepad_id)) {
+        return;
+    }
 
-        // exit_on_error: false (a gamepad open failure should not exit scrcpy)
-        if (!sc_aoa_push_open(gamepad->aoa, &hid_open, false)) {
-            LOGW("Could not push AOA HID open (gamepad)");
-        }
-    } else {
-        assert(event->type == SC_GAMEPAD_DEVICE_REMOVED);
+    // exit_on_error: false (a gamepad open failure should not exit scrcpy)
+    if (!sc_aoa_push_open(gamepad->aoa, &hid_open, false)) {
+        LOGW("Could not push AOA HID open (gamepad)");
+    }
+}
 
-        struct sc_hid_close hid_close;
-        if (!sc_hid_gamepad_generate_close(&gamepad->hid, &hid_close,
-                                           event->gamepad_id)) {
-            return;
-        }
+static void
+sc_gamepad_processor_process_gamepad_removed(struct sc_gamepad_processor *gp,
+                                const struct sc_gamepad_device_event *event) {
+    struct sc_gamepad_aoa *gamepad = DOWNCAST(gp);
 
-        if (!sc_aoa_push_close(gamepad->aoa, &hid_close)) {
-            LOGW("Could not push AOA HID close (gamepad)");
-        }
+    struct sc_hid_close hid_close;
+    if (!sc_hid_gamepad_generate_close(&gamepad->hid, &hid_close,
+                                       event->gamepad_id)) {
+        return;
+    }
+
+    if (!sc_aoa_push_close(gamepad->aoa, &hid_close)) {
+        LOGW("Could not push AOA HID close (gamepad)");
     }
 }
 
@@ -76,7 +78,8 @@ sc_gamepad_aoa_init(struct sc_gamepad_aoa *gamepad, struct sc_aoa *aoa) {
     sc_hid_gamepad_init(&gamepad->hid);
 
     static const struct sc_gamepad_processor_ops ops = {
-        .process_gamepad_device = sc_gamepad_processor_process_gamepad_device,
+        .process_gamepad_added = sc_gamepad_processor_process_gamepad_added,
+        .process_gamepad_removed = sc_gamepad_processor_process_gamepad_removed,
         .process_gamepad_axis = sc_gamepad_processor_process_gamepad_axis,
         .process_gamepad_button = sc_gamepad_processor_process_gamepad_button,
     };

--- a/app/src/usb/screen_otg.c
+++ b/app/src/usb/screen_otg.c
@@ -175,7 +175,6 @@ sc_screen_otg_process_gamepad_device(struct sc_screen_otg *screen,
     assert(screen->gamepad);
     struct sc_gamepad_processor *gp = &screen->gamepad->gamepad_processor;
 
-    SDL_JoystickID id;
     if (event->type == SDL_CONTROLLERDEVICEADDED) {
         SDL_GameController *gc = SDL_GameControllerOpen(event->which);
         if (!gc) {
@@ -190,9 +189,12 @@ sc_screen_otg_process_gamepad_device(struct sc_screen_otg *screen,
             return;
         }
 
-        id = SDL_JoystickInstanceID(joystick);
+        struct sc_gamepad_device_event evt = {
+            .gamepad_id = SDL_JoystickInstanceID(joystick),
+        };
+        gp->ops->process_gamepad_added(gp, &evt);
     } else if (event->type == SDL_CONTROLLERDEVICEREMOVED) {
-        id = event->which;
+        SDL_JoystickID id = event->which;
 
         SDL_GameController *gc = SDL_GameControllerFromInstanceID(id);
         if (gc) {
@@ -200,16 +202,12 @@ sc_screen_otg_process_gamepad_device(struct sc_screen_otg *screen,
         } else {
             LOGW("Unknown gamepad device removed");
         }
-    } else {
-        // Nothing to do
-        return;
-    }
 
-    struct sc_gamepad_device_event evt = {
-        .type = sc_gamepad_device_event_type_from_sdl_type(event->type),
-        .gamepad_id = id,
-    };
-    gp->ops->process_gamepad_device(gp, &evt);
+        struct sc_gamepad_device_event evt = {
+            .gamepad_id = id,
+        };
+        gp->ops->process_gamepad_removed(gp, &evt);
+    }
 }
 
 static void

--- a/app/tests/test_control_msg_serialize.c
+++ b/app/tests/test_control_msg_serialize.c
@@ -329,6 +329,8 @@ static void test_serialize_uhid_create(void) {
         .type = SC_CONTROL_MSG_TYPE_UHID_CREATE,
         .uhid_create = {
             .id = 42,
+            .vendor_id = 0x1234,
+            .product_id = 0x5678,
             .name = "ABC",
             .report_desc_size = sizeof(report_desc),
             .report_desc = report_desc,
@@ -337,11 +339,13 @@ static void test_serialize_uhid_create(void) {
 
     uint8_t buf[SC_CONTROL_MSG_MAX_SIZE];
     size_t size = sc_control_msg_serialize(&msg, buf);
-    assert(size == 20);
+    assert(size == 24);
 
     const uint8_t expected[] = {
         SC_CONTROL_MSG_TYPE_UHID_CREATE,
         0, 42, // id
+        0x12, 0x34, // vendor id
+        0x56, 0x78, // product id
         3, // name size
         65, 66, 67, // "ABC"
         0, 11, // report desc size

--- a/server/src/main/java/com/genymobile/scrcpy/control/ControlMessage.java
+++ b/server/src/main/java/com/genymobile/scrcpy/control/ControlMessage.java
@@ -51,6 +51,8 @@ public final class ControlMessage {
     private int id;
     private byte[] data;
     private boolean on;
+    private int vendorId;
+    private int productId;
 
     private ControlMessage() {
     }
@@ -131,10 +133,12 @@ public final class ControlMessage {
         return msg;
     }
 
-    public static ControlMessage createUhidCreate(int id, String name, byte[] reportDesc) {
+    public static ControlMessage createUhidCreate(int id, int vendorId, int productId, String name, byte[] reportDesc) {
         ControlMessage msg = new ControlMessage();
         msg.type = TYPE_UHID_CREATE;
         msg.id = id;
+        msg.vendorId = vendorId;
+        msg.productId = productId;
         msg.text = name;
         msg.data = reportDesc;
         return msg;
@@ -236,5 +240,13 @@ public final class ControlMessage {
 
     public boolean getOn() {
         return on;
+    }
+
+    public int getVendorId() {
+        return vendorId;
+    }
+
+    public int getProductId() {
+        return productId;
     }
 }

--- a/server/src/main/java/com/genymobile/scrcpy/control/ControlMessageReader.java
+++ b/server/src/main/java/com/genymobile/scrcpy/control/ControlMessageReader.java
@@ -142,9 +142,11 @@ public class ControlMessageReader {
 
     private ControlMessage parseUhidCreate() throws IOException {
         int id = dis.readUnsignedShort();
+        int vendorId = dis.readUnsignedShort();
+        int productId = dis.readUnsignedShort();
         String name = parseString(1);
         byte[] data = parseByteArray(2);
-        return ControlMessage.createUhidCreate(id, name, data);
+        return ControlMessage.createUhidCreate(id, vendorId, productId, name, data);
     }
 
     private ControlMessage parseUhidInput() throws IOException {

--- a/server/src/main/java/com/genymobile/scrcpy/control/Controller.java
+++ b/server/src/main/java/com/genymobile/scrcpy/control/Controller.java
@@ -290,7 +290,7 @@ public class Controller implements AsyncProcessor, VirtualDisplayListener {
                 Device.rotateDevice(getActionDisplayId());
                 break;
             case ControlMessage.TYPE_UHID_CREATE:
-                getUhidManager().open(msg.getId(), msg.getText(), msg.getData());
+                getUhidManager().open(msg.getId(), msg.getVendorId(), msg.getProductId(), msg.getText(), msg.getData());
                 break;
             case ControlMessage.TYPE_UHID_INPUT:
                 getUhidManager().writeInput(msg.getId(), msg.getData());

--- a/server/src/main/java/com/genymobile/scrcpy/control/UhidManager.java
+++ b/server/src/main/java/com/genymobile/scrcpy/control/UhidManager.java
@@ -48,7 +48,7 @@ public final class UhidManager {
         }
     }
 
-    public void open(int id, String name, byte[] reportDesc) throws IOException {
+    public void open(int id, int vendorId, int productId, String name, byte[] reportDesc) throws IOException {
         try {
             FileDescriptor fd = Os.open("/dev/uhid", OsConstants.O_RDWR, 0);
             try {
@@ -58,7 +58,7 @@ public final class UhidManager {
                     close(old);
                 }
 
-                byte[] req = buildUhidCreate2Req(name, reportDesc);
+                byte[] req = buildUhidCreate2Req(vendorId, productId, name, reportDesc);
                 Os.write(fd, req, 0, req.length);
 
                 registerUhidListener(id, fd);
@@ -148,7 +148,7 @@ public final class UhidManager {
         }
     }
 
-    private static byte[] buildUhidCreate2Req(String name, byte[] reportDesc) {
+    private static byte[] buildUhidCreate2Req(int vendorId, int productId, String name, byte[] reportDesc) {
         /*
          * struct uhid_event {
          *     uint32_t type;
@@ -183,8 +183,8 @@ public final class UhidManager {
 
         buf.putShort((short) reportDesc.length);
         buf.putShort(BUS_VIRTUAL);
-        buf.putInt(0); // vendor id
-        buf.putInt(0); // product id
+        buf.putInt(vendorId);
+        buf.putInt(productId);
         buf.putInt(0); // version
         buf.putInt(0); // country;
         buf.put(reportDesc);

--- a/server/src/main/java/com/genymobile/scrcpy/control/UhidManager.java
+++ b/server/src/main/java/com/genymobile/scrcpy/control/UhidManager.java
@@ -174,7 +174,7 @@ public final class UhidManager {
         ByteBuffer buf = ByteBuffer.allocate(280 + reportDesc.length).order(ByteOrder.nativeOrder());
         buf.putInt(UHID_CREATE2);
 
-        String actualName = name.isEmpty() ? "scrcpy" : "scrcpy: " + name;
+        String actualName = name.isEmpty() ? "scrcpy" : name;
         byte[] utf8Name = actualName.getBytes(StandardCharsets.UTF_8);
         int len = StringUtils.getUtf8TruncationIndex(utf8Name, 127);
         assert len <= 127;

--- a/server/src/test/java/com/genymobile/scrcpy/control/ControlMessageReaderTest.java
+++ b/server/src/test/java/com/genymobile/scrcpy/control/ControlMessageReaderTest.java
@@ -322,6 +322,8 @@ public class ControlMessageReaderTest {
         DataOutputStream dos = new DataOutputStream(bos);
         dos.writeByte(ControlMessage.TYPE_UHID_CREATE);
         dos.writeShort(42); // id
+        dos.writeShort(0x1234); // vendorId
+        dos.writeShort(0x5678); // productId
         dos.writeByte(3); // name size
         dos.write("ABC".getBytes(StandardCharsets.US_ASCII));
         byte[] data = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11};
@@ -335,6 +337,8 @@ public class ControlMessageReaderTest {
         ControlMessage event = reader.read();
         Assert.assertEquals(ControlMessage.TYPE_UHID_CREATE, event.getType());
         Assert.assertEquals(42, event.getId());
+        Assert.assertEquals(0x1234, event.getVendorId());
+        Assert.assertEquals(0x5678, event.getProductId());
         Assert.assertEquals("ABC", event.getText());
         Assert.assertArrayEquals(data, event.getData());
 

